design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+
+### Goals
+
+1. To provide the interface that will make possible to apply different remediation strategy for each cloud or bare metal providers.
+2. To provide details regarding the progress of the remediation operation.
+
+### Non-Goals
+
+1. To provide a remediation strategy for each cloud or bare metal providers.
+2. To provide the integration with the **MachineHealthCheck** controller.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+It's common that high available environment should have a mechanism to restore unhealthy nodes and when delete and create a new node looks feasible solution for a cloud provider, in case of bare metal hosts with additional infrastructure configuration it looks like too destructive approach.
+To make possible to specify the remediation strategy for different failure cases and for nodes that run on different providers, this proposal introduces **MachineRemediationRequest** controller that should monitor all **MachineRemediationRequest** CR's and apply the remediation strategy according to details specified under the CR.
+When the **MachineRemediationRequest** will apply the remediation strategy, one who is initiating it will be the **MachineHealthCheck** controller. It will monitor nodes, and for each unhealthy node will create **MachineRemediationRequest** CR, if after the specified timeout the **MachineRemediationRequest** will not have the `remediationState` equal to `Succeeded`, the **MachineHealthCheck** controller will notify about the remediation failure, and can try to use different type of the remediation strategy.
+
+#### MachineRemediationRequest spec API
+
+This proposal introduces a new API type: **MachineRemediationRequest**, that makes possible to create simple **MachineRemediationRequest** object with the type of the remediation strategy for the specific machine.
+
+##### Example
+
+```yaml
+apiVersion: machineremediationrequest.metal3.io/v1alpha1
+kind: MachineRemediationRequest
+metadata:
+  name: mrr-test
+  namespace: openshift-machine-api
+spec:
+  type: reboot
+  machineName: test-machine
+```
+
+#### MachineRemediationRequest status API
+
+**MachineRemediationRequest** status will show what the state the remediation operation has and the time when the remediation operation started.
+
+```yaml
+status:
+  remediationState: InProgress
+  remediationStartTime: Thu, 20 Jun 2019 03:38:39 -0400
+```
+
+### Risks and Mitigations
+
+It can introduce some integration complexity between **MachineHealthCheck** and **MachineRemediationRequest** controllers,
+that will make harder to debug and to test things.
+
+## Design Details
+
+### Work Items
+
+* Create new **MachineRemediationRequest** CRD.
+* Write a new controller that will watch **MachineRemediationRequest** CR's.
+* Write functional tests under <https://github.com/openshift/cluster-api-actuator-pkg> repository.
+* Create an operator to deploy the **MachineRemediationRequest** controller.
+* Create integration with the OLM.
```

Comment:
This step is also specific to openshift.

Reply:
You can deploy OLM on top of k8s cluster so I prefer to leave it here